File: Haskell/SnocList.hs
module SnocList (SnocList(SL),SnocList(Null),peek,dequeue,enqueue,isEmpty,showQueue) where

-- This module is an implementation of a Queue

import Data.Bool

-- data structure for SnocLists
-- each one could either be a SnocList or an empty list (Null)
data SnocList = SL Head SnocList | Null deriving(Show,Eq)
type Head = Char
 
-- returns the current head of the list (next in queue)
peek :: SnocList -> Head
peek (SL h s) = h

-- removes the head of the list
dequeue :: SnocList -> SnocList
dequeue (SL h s) = s

-- adds a character to the end of the list
enqueue :: Char -> SnocList -> SnocList
enqueue c Null = SL c Null
enqueue c (SL h p) = SL c (SL h p)

-- only true if list is empty
isEmpty :: SnocList -> Bool
isEmpty Null = True
isEmpty s = False

-- shows SnocList as a String = [Char]
showQueue :: SnocList -> String
showQueue Null = ""
showQueue q = showHelper q

-- showHelper turns a SnocList into a list of Chars
-- It displays as a string because String = [Char]
showHelper :: SnocList -> [Char]
showHelper (SL h Null) = [h]
showHelper (SL h q) = showHelper q ++ [h]

File: Java/SnocList.java
import java.io.*;
import java.lang.ArrayIndexOutOfBoundsException;
import java.lang.Character;
import java.lang.Exception;
import java.lang.Throwable;
import java.util.ArrayList;

public class SnocList implements Queue{

    private char c ;

    private SnocList l ;

    SnocList(char c, SnocList l) {
        this.c = c;
        this.l = l;
    }

    public char peek(){
    // front
        return this.c;
    }

    public void dequeue(){
    // front
        this = l;
        this.c = "\0"
    }

    public void enqueue(char ch){
    // back
        if (this.l.c == "\0") this.l.c = ch;
        else this.l.enqueue(ch);
    }


    public boolean isEmpty(){
        if (this.c != '\0') return true;
        else return false;
    }
}